### PR TITLE
PP-6268 Implement Service Switcher

### PR DESCRIPTION
**Pull request by rorymalcolm**
- Implements the service switcher, allowing routes to provide a
serviceId through the serviceSwitcher middleware, which will then
redirect users to that service before loading the page
- Tests to ensure that functionality works as expected.

### CODE REVIEW

File: app/middleware/service-switcher.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+const logger = require('../utils/logger')(__filename)
+const validAccountId = require('../utils/valid_account_id.js')
+const { renderErrorView } = require('../utils/response.js')
+
+const notAuthorised = (req, res) => {
+  return renderErrorView(req, res, 'You do not have the rights to access this service.', 403)
+}
+
+module.exports = (req, res, next) => {
+  let newAccountId = req.params.serviceId
```
Comment (stephencdaly) [inline app/middleware/service-switcher.js]: Does the path contain a service id or a gateway account id? I'm confused.
Also this should be a const.

File: app/routes.js [hunk 2]
```diff
@@ -223,6 +224,7 @@ module.exports.bind = function (app) {
   app.get(transactions.download, xraySegmentCls, permission('transactions-download:read'), getAccount, paymentMethodIsCard, transactionsDownloadController)
   app.get(transactions.detail, xraySegmentCls, permission('transactions-details:read'), getAccount, paymentMethodIsCard, transactionDetailController)
   app.post(transactions.refund, xraySegmentCls, permission('refunds:create'), getAccount, paymentMethodIsCard, transactionRefundController)
+  app.get(transactions.serviceSwitchDetail, xraySegmentCls, permission('transactions-details:read'), switchService, getAccount, paymentMethodIsCard, transactionDetailController)
```
Comment (stephencdaly) [inline app/routes.js]: I think the `permission` middleware should come after the `switchService` middleware, as we want to check the user has the permission for the service we're switching to.

File: test/integration/service_switcher_ft_tests.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+      })
+  })
+
+  describe('The transaction view endpoint', function () {
+    it('should return the correct transaction view through the service switching middleware', function (done) {
+      let transactionId = 888
+      ledgerMock.get(ledgerTransactionPathFor(transactionId))
+        .reply(200,
+          {
+            'gateway_account_id': gatewayAccountId,
```
Comment (stephencdaly) [inline test/integration/service_switcher_ft_tests.js]: can you use a fixture from `ledger_transaction_fixtures` to produce this response.

File: test/integration/service_switcher_ft_tests.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    let user = session.getUser({
+      gateway_account_ids: [gatewayAccountId],
+      permissions: [{ name: permissions }]
+    })
+    app = session.getAppWithLoggedInUser(getApp(), user)
+
+    userCreator.mockUserResponse(user.toJson(), done)
+    connectorMock.get(CONNECTOR_ACCOUNT_PATH)
+      .reply(200, {
+        'payment_provider': 'sandbox',
```
Comment (stephencdaly) [inline test/integration/service_switcher_ft_tests.js]: can you use a fixture from `gateway_account_fixtures` to construct this response.

File: app/services/auth_service.js [hunk 5]
```diff
@@ -145,12 +145,6 @@ function getCurrentGatewayAccountId (req) {
     logger.error(`Could not resolve the gatewayAccountId for user: ${lodash.get(req, 'user.externalId')}`)
     return null
   }
-  // check if we don't have Cookie value
-  // or if it's different user  / different userGatewayAccountIds
-  if ((!currentGatewayAccountId) ||
-    (currentServiceGatewayAccountIds.indexOf(currentGatewayAccountId) === -1)) {
-    currentGatewayAccountId = currentServiceGatewayAccountIds[0]
-  }
```
Comment (stephencdaly) [inline app/services/auth_service.js]: do we not still need this behaviour in other circumstances?

File: app/middleware/service-switcher.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  if (validAccountId(newAccountId, req.user)) {
```
Comment (stephencdaly) [inline app/middleware/service-switcher.js]: Could this be handled by the `permission` middleware if we change the ordering?

File: app/routes.js (hunk 2 above)
Comment (rorymalcolm) [reply app/routes.js]: oops

**Pull request merged**